internal/cmd/gocdk/launch.go

```diff
@@ -187,6 +209,172 @@ func (local *localLauncher) Launch(ctx context.Context, input *LaunchInput) (*ur
 	return serveURL, nil
 }
 
+// cloudRunLauncher pushes Docker containers to Google Container Registry and
+// creates/updates a Cloud Run service.
+type cloudRunLauncher struct {
+	logger    *log.Logger
+	client    *cloudrun.APIService
+	dockerEnv []string
+	dockerDir string
+}
+
+// Launch implements Launcher.Launch.
+func (crl *cloudRunLauncher) Launch(ctx context.Context, input *LaunchInput) (*url.URL, error) {
+	projectID := specifierStringValue(input.Specifier, "project_id")
+	location := specifierStringValue(input.Specifier, "location")
+	serviceName := specifierStringValue(input.Specifier, "service_name")
+	if projectID == "" || location == "" || serviceName == "" {
+		return nil, xerrors.New("cloud run launch: specifier does include project_id, location, and/or service_name")
+	}
+
+	// Push to GCR if needed.
+	// TODO(light): Check for presence in Docker daemon and country-specific
+	// prefixes.
+	if strings.HasPrefix(input.DockerImage, "gcr.io/") {
+		if err := crl.push(ctx, input.DockerImage); err != nil {
+			return nil, xerrors.Errorf("cloud run launch: %w", err)
+		}
+	}
+
+	// Launch on Cloud Run.
+	var env []*cloudrun.EnvVar
+	for i, inputVar := range input.Env {
+		eqIdx := strings.IndexByte(inputVar, '=')
+		if eqIdx == -1 {
+			return nil, xerrors.Errorf("cloud run launch: bad format for specifier env[%d] (got %q)", i, inputVar)
+		}
+		env = append(env, &cloudrun.EnvVar{
+			Name:  inputVar[:eqIdx],
+			Value: inputVar[eqIdx+1:],
+		})
+	}
+	serviceMeta := &cloudrun.ObjectMeta{
+		Name:      serviceName,
+		Namespace: projectID,
+	}
+	serviceSpec := &cloudrun.ServiceSpec{
+		RunLatest: &cloudrun.ServiceSpecRunLatest{
+			Configuration: &cloudrun.ConfigurationSpec{
+				RevisionTemplate: &cloudrun.RevisionTemplate{
+					Spec: &cloudrun.RevisionSpec{
+						Container: &cloudrun.Container{
+							Image: input.DockerImage,
+							Env:   env,
+							// TODO(light): Add liveness and readiness probes.
+						},
+						ContainerConcurrency: 0, // thread-safe
+					},
+				},
+			},
+		},
+	}
+	locationString := "projects/" + projectID + "/locations/" + location
+	serviceString := locationString + "/services/" + serviceName
+	createCall := crl.client.Projects.Locations.Services.Create(locationString, &cloudrun.Service{
+		ApiVersion: "serving.knative.dev/v1alpha1",
+		Kind:       "Service",
+		Metadata:   serviceMeta,
+		Spec:       serviceSpec,
+	})
+	_, err := createCall.Context(ctx).Do()
+	if apiError := (*googleapi.Error)(nil); xerrors.As(err, &apiError) && apiError.Code == http.StatusConflict {
+		// Already exists, add revision.
+		replaceCall := crl.client.Projects.Locations.Services.ReplaceService(serviceString, &cloudrun.Service{
+			ApiVersion: "serving.knative.dev/v1alpha1",
+			Kind:       "Service",
+			Metadata:   serviceMeta,
+			Spec:       serviceSpec,
+		})
+		// Handle error below in the same way a create call is handled.
+		_, err = replaceCall.Context(ctx).Do()
+	} else if err == nil {
```

Comment:
Nit: I found this error flow a bit hard to follow. How about flipping the if/else?

```
err := createCall.Do()
if err == nil {
  // set IAM
  err = policyCall.Do()
} else if err == StatusConflict {
  // Add revision
  err = replaceCall.Do()
}

if err != nil {
  // either the create failed, or the policy call failed, or the replace failed
  return err
}

...
```